File: ui/ai.js
const tf = require('@tensorflow/tfjs-node');
const fs = require('fs');
const fetch = require('node-fetch');
const backend = require('./backend/Backend');


const encoderModelPath = 'ai/encoder_model.keras';
const decoderModelPath = 'ai/decoder_model.keras';
const EOS = 0;
let SOS;

backend.getNumberOfSigns().then(numberOfSigns => {
    SOS = parseInt(numberOfSigns) + 1;
    // Rest of your code that depends on SOS goes here
});
const encoderModel = await tf.loadLayersModel('file://' + encoderModelPath);
const decoderModel = await tf.loadLayersModel('file://' + decoderModelPath);


    

// Now you can use these models for inference
function decodeSequence(inputSeq) {
    // Encode the input as state vectors.
    const statesValue = encoderModel.predict(inputSeq);

    // Generate empty target sequence of length 1.
    const targetSeq = tf.zeros([1, 1, 1]);
    // Populate the first character of target sequence with the start character.
    targetSeq.tensor[0, 0, 0] = SOS;

    // Sampling loop for a batch of sequences
    const decodedSequence = [];
    let lastTokenIndex = null;
    for (let i = 0; i < 20; i++) {
        const [outputTokens, h, c] = decoderModel.predict([targetSeq].concat(statesValue));

        // Sample a token
        let sampledTokenIndex = tf.argMax(outputTokens.tensor[0, -1]);
        sampledTokenIndex = sampledTokenIndex.dataSync()[0];
        if (sampledTokenIndex === lastTokenIndex) {
            // If the model tries to predict the same sign again, choose the sign with the second highest probability instead
            const topTokens = tf.topk(outputTokens.tensor[0, -1], 2);
            const topTokensData = topTokens.indices.dataSync();
            sampledTokenIndex = topTokensData[0] === lastTokenIndex ? topTokensData[1] : topTokensData[0];
        }
        if (sampledTokenIndex === EOS) {
            break;
        }
        decodedSequence.push(sampledTokenIndex);

        // Update the target sequence (of length 1).
        targetSeq.tensor[0, 0, 0] = sampledTokenIndex;

        // Update states
        statesValue[0].dispose();
        statesValue[1].dispose();
        statesValue[0] = h;
        statesValue[1] = c;

        // Remember the last predicted token
        lastTokenIndex = sampledTokenIndex;
    }

    return decodedSequence;
}



// Usage example
const inputSeq = tf.tensor(); // Replace with your input sequence
const decodeSequence = await loadModels();
const decodedSeq = decodeSequence(inputSeq);
console.log(decodedSeq);

File: ui/ai.py
import numpy as np
import tensorflow as tf
import keras


encoder_model = keras.models.load_model('models/encoder_model.keras')
decoder_model = keras.models.load_model('models/decoder_model.keras')

# Define start and end tokens
SOS = 7352 # Start of sequence
EOS = 0 # End of sequence

# Define a function to predict decoder output
@tf.function(reduce_retracing=True)
def predict_decoder(decoder_inputs, decoder_states, encoder_outputs):
    return decoder_model([decoder_inputs] + decoder_states + [encoder_outputs])

def decode_sequence(input_seq):

    # Reshape input data
    pose_input_data = np.array(input_seq[0]).reshape(1, -1, 23 * 3).astype(np.float16)
    left_hand_input_data = np.array(input_seq[1]).reshape(1, -1, 21 * 3).astype(np.float16)
    right_hand_input_data = np.array(input_seq[2]).reshape(1, -1, 21 * 3).astype(np.float16)

    # Encode the input sequence to get the encoder states
    encoder_outputs, state_h, state_c = encoder_model.predict([pose_input_data, left_hand_input_data, right_hand_input_data])
    
    # Save the states value
    states_value = [state_h, state_c]

    # Start the sequence with the start token
    target_seq = np.array([[SOS]]).reshape(1, -1, 1).astype(np.int16) 

    # Initialize the decoded sequence
    decoded_sequence = [] 

    # Initialize the last token index
    last_token_index = None
    for i in range(26): # Maximum sequence length, the alphabet is the largest sequence it is trained on

        # Predict the output token
        output_tokens, h, c = predict_decoder(target_seq, states_value, encoder_outputs)

        # Sample a token
        sampled_token_index = np.argmax(output_tokens[0, -1, :])

        # If the sampled token is the same as the last token, sample the second highest token
        if sampled_token_index == last_token_index:
            top_tokens = np.argsort(output_tokens[0, -1, :])[-2:]
            sampled_token_index = top_tokens[0] if top_tokens[1] == last_token_index else top_tokens[1]

        # Break if the sampled token is the end token
        if sampled_token_index == EOS:
            break

        # Append the sampled token to the decoded sequence
        decoded_sequence.append(sampled_token_index)

        # Update the target sequence
        target_seq = np.append(target_seq, [[[sampled_token_index]]], axis=1)

        # Update states value
        states_value = [h, c]

        # Update the last token index
        last_token_index = sampled_token_index

    return decoded_sequence


